### DISCUSSION TITLE
On-premise databases: How should I deploy my application, without a bundled SQLite database? I'm confused

### DISCUSSION THREAD

**Discussion by birgersp** [prisma/prisma#21482] [Q&A] [answered]
### Question

I am making an application that uses prisma and SQLite. The application may use an existing database (if present). And then my idea was that if there was no database, the application should just create one.

I'm confused how to do this. FWIW my application is a docker image. But even if it wasn't, I'd feel awkward having to bundle the "empty" (except the tables created etc by prisma migrate of course) database with the application. And what if I wanted to "clear" the database, it would be easy to just remove it and have my application create a new one.

I tried copying the migration scripts into my application, and then run code at the start to check if the database had the tables etc and try to apply the scripts if it didnt. First of all it didn't work and even if it did it feels very hacky to have to do this.

Perhaps there should be a `PrismaClient.createSqliteDb` or something that would create the db?

Hoping someone can give me some tips or join a discussion about this.

**Comment by ludralph** [comment]
Hi @birgersp 👋 

When you run a Prisma migration, it checks for the existence of the database file and creates it if it doesn't exist. This behavior is particularly useful when working with SQLite databases.
when you run the command:
```
npx prisma migrate dev --name init
```
It does the following:

1. Creates a new SQL migration file for the migration in the prisma/migrations directory.
2. Executes the SQL migration file against the database.
3. Runs `prisma generate` under the hood, which installs the `@prisma/client` package and generates a tailored Prisma Client API based on your models.

If the SQLite database file does not exist before running this command, it also creates it inside the prisma directory with the name `dev.db` as defined via the environment variable in the `.env` file.

You should note that Prisma always connects to the database first, regardless of the command. Connecting to a non-existent SQLite database has the side effect of creating it. See [issue comment](https://github.com/prisma/prisma/issues/7969#issuecomment-872199266)

Therefore, you don't need to bundle an empty database with your application or manually run migration scripts. Prisma Migrate will handle the creation of the database and apply the necessary schema changes when you run the migration command.

If you want to `clear` the database, you can simply delete the SQLite file, and Prisma Migrate will create a new one the next time it runs.

For your Docker setup, you can include the Prisma migration command in your Dockerfile or entrypoint script to ensure that the database is created and up to date when the container starts.

Remember to include the `.env` file in your Docker image or provide the necessary environment variables to specify the database connection string.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by birgersp** [reply]
Thanks for a great response! But one thing is unclear to me.

If there already is a database, and I run "npx prisma migrate dev --name init", what happens exactly?

What if the database doesn't match the prisma schema? Will I risk deleting or damaging the database somehow?

https://www.prisma.io/docs/orm/reference/prisma-cli-reference#migrate-dev

Also, the docs states that "prisma migrate dev" is "for use in development environments only". So tbh I do feel like database-"initialization" (specficially, creation of a SQLite database (if none exists) that matches the prisma schema) is something that is lacking in Prisma.

**Reply by janpio** [reply]
You can just use `prisma migrate deploy` in your production environment to run the migration files that you created during development. If no database exists, it will be created. If a database already exists, it will apply the missing migration files. If all migration files are already applied, it does nothing.

With SQLite "production environment" is a bit tricky to think about, as it is not just a different remote database. But it could for example be your production/main branch where the _real_ database is included. That's the "downside" of having a database as a file.

**Reply by birgersp** [reply]
I just realized why running prisma commands in the docker container isn't a viable option; it would require us to have prisma (not just the client) installed alongside the application. The size of our docker image would increase by 5x.

I appreciate your answers, but I think my inital question didn't really clarify what I am trying to achieve here.

So to update and clarify my initial post:

* Prisma will only run if the database matches the prisma schema.
* I'd rather not run any Prisma CLI commands in the production environment.

Current workaround:
When the application starts, we use a ["sqlite3"](https://www.npmjs.com/package/sqlite3) client to create the db (if none exists) and automatically run necessary migration steps on an existing database.
Then (after the previous step succeedes) create the PrismaClient instance and use Prisma as usual.

**Reply by janpio** [reply, marked as answer]
Yes, your workaround sounds sensible then. Because for SQLite "I'd rather not run any Prisma CLI commands in the production environment." is just not a supported workflow right now, except of course if you migrate _outside_ production and then copy the database file there (or make it accessible). I would assume that this is what many SQLite users are doing.

**Reply by birgersp** [reply]
Thats fair, just something to keep in mind for people in the same situation as me. Hopefully some could find this discussion helpful.
